lib/charms/mongodb/v1/mongodb_tls.py

```diff
@@ -302,21 +315,93 @@ def _on_certificate_expiring(self, event: CertificateExpiringEvent) -> None:
 
         self.set_tls_secret(internal, Config.TLS.SECRET_CSR_LABEL, new_csr.decode("utf-8"))
 
-    def _get_sans(self) -> List[str]:
+    def get_new_sans(self) -> Dict:
         """Create a list of DNS names for a MongoDB unit.
 
         Returns:
             A list representing the hostnames of the MongoDB unit.
         """
+        sans = {}
+
         unit_id = self.charm.unit.name.split("/")[1]
-        return [
+        sans[SANS_DNS_KEY] = [
             f"{self.charm.app.name}-{unit_id}",
             socket.getfqdn(),
-            f"{self.charm.app.name}-{unit_id}.{self.charm.app.name}-endpoints",
-            str(self.charm.model.get_binding(self.peer_relation).network.bind_address),
             "localhost",
+            f"{self.charm.app.name}-{unit_id}.{self.charm.app.name}-endpoints",
+        ]
+
+        sans[SANS_IPS_KEY] = [
+            str(self.charm.model.get_binding(self.peer_relation).network.bind_address)
         ]
 
+        if self.charm.is_role(Config.Role.MONGOS) and self.charm.is_external_client:
+            sans[SANS_IPS_KEY].append(
+                self.charm.get_ext_mongos_host(self.charm.unit, incl_port=False)
+            )
+
+        return sans
+
+    def get_current_sans(self, internal: bool) -> List | None:
+        """Gets the current SANs for the unit cert."""
+        # if unit has no certificates do not proceed.
+        if not self.is_tls_enabled(internal=internal):
+            return
+
+        try:
+            sans_lines = self.get_sans_from_host(internal)
+        except (subprocess.CalledProcessError, ExecError) as e:
+            logger.error(e.stdout)
+            raise e
+
+        line = ""
+        for line in sans_lines:
+            if "DNS" in line or "IP" in line:
+                break
+
+        sans_ip = []
+        sans_dns = []
+        for item in line.split(", "):
+            if ":" not in item:
+                continue
+
+            san_type, san_value = item.split(":")
+
+            if san_type == "DNS":
+                sans_dns.append(san_value)
+            if san_type == "IP Address":
+                sans_ip.append(san_value)
+
+        return {SANS_IPS_KEY: sorted(sans_ip), SANS_DNS_KEY: sorted(sans_dns)}
+
+    def get_sans_from_host(self, internal) -> List[str]:
+        """Returns the sans lines from the host.
+
+        Raises: subprocess.CalledProcessError, ExecError
+        """
+        pem_file = Config.TLS.INT_PEM_FILE if internal else Config.TLS.EXT_PEM_FILE
+
+        command = [
```

Comment:
I would go with something along the lines of that:
```python3
from cryptography import x509
from cryptography.hazmat.backends import default_backend

def get_current_sans(self, internal: bool) -> dict[str, str]:
    [Some code to get the certificate data in bytes]
    try:
        cert = cert = x509.load_pem_x509_certificate(data, default_backend())
        sans = cert.extensions.get_extenstion_for_class(x509.SubjectAlternativeName).value
        sans_ip = [str(san) for san in sans.get_values_for_type(x509.IPAddress)]
        sans_dns = [str(san) for san in sans.get_values_for_type(x509.DNSName)]
    except x509.ExtensionNotFound:
        sans_ip = []
        sans_dns = []
    return {SANS_IPS_KEY: sorted(sans_ip), SANS_DNS_KEY: sorted(sans_dns)}
```
Just need to provide the code that gets the certs data (so either from the workload with a cat or get from secrets) and this will be easier, cleaner, no risk of text interface change, and no need to ensure that openssl binary is installed on the workload

Reply:
If this works out of the box then lets go with it - otherwise lets keep it as is (we use the same function here as Kafka - so there is value in that)